src/exo/cursors.py

```diff
@@ -0,0 +1,228 @@
+from __future__ import annotations
+
+import weakref
+from dataclasses import dataclass
+from enum import Enum
+from typing import Union, List, Iterable
+from weakref import ReferenceType
+
+from .API_types import ProcedureBase
+from .LoopIR import LoopIR
+
+
+class CursorKind(Enum):
+    Node = 0  # Pointing directly at a node
+    # BlockFront = 1  # Gap before the first statement of a block
+    # BlockEnd = 2  # Gap after the last statement of a block
+    GapBefore = 3  # Gap before the node in a block
+    GapAfter  = 4  # Gap after the node in a block
+
+
+class ExoIR(Enum):
+    Assign = 0
+    Reduce = 1
+    WriteConfig = 2
+    Pass = 3
+    If = 4
+    For = 5
+    Alloc = 6
+    Call = 7
+    WindowStmt = 8
+
+
+class InvalidCursorError(Exception):
+    pass
+
+
+@dataclass
+class Cursor:
+    _proc: ReferenceType[ProcedureBase]
+    _node: ReferenceType[Union[LoopIR.proc, LoopIR.stmt, LoopIR.expr]]
```

Comment:
Why should a cursor ever be pointed at a `proc` rather than its body?

Reply:
Secondly, looking at this implementation it seems like we only ever really need to have a `LoopIR.proc`, not a `Procedure` (which is the external wrapper object for the API)  Should we change what's stored here then?